### ISSUE TITLE
java.security.NoSuchAlgorithmException: SunX509 KeyManagerFactory not available

### ISSUE THREAD

**Issue by vineetravi107** [drallgood/jpasskit#94]
Now i am getting the below error

![capture](https://user-images.githubusercontent.com/9457234/29814074-b999d48e-8cc9-11e7-9dd9-272080aab5ed.JPG)

I am using the following jar versions on Oracle WebLogic Server JDK1.8 IBM AIX Linix.

![capture1](https://user-images.githubusercontent.com/9457234/29814160-f59b467a-8cc9-11e7-87e2-776f4a7c6b86.JPG)

**Comment by vineetravi107** [comment]
Is this error because we are using IBM AIX sever? Will this JAR not support this?

**Comment by drallgood** [comment]
Very likely yes. 
I really don't know how you could fix this. Maybe you can configure another key manager by providing the `ssl.KeyManagerFactory.algorithm` environment variable
See also this: 
- https://coderanch.com/t/133186/Algorithm
- https://stackoverflow.com/questions/9926349/sunx509-trustmanagerfactory-not-available-on-aix-ssl-webservice-client

**Comment by vineetravi107** [comment]
Is it possible if you can provide updated jPassKit JAR to support  Oracle WebLogic Platform on IBM AIX ?

**Comment by drallgood** [comment]
Sorry, I don't think I'm able to do that. I don't have the environment to test this plus I don't think this is something I can even control.
Have you tried the things mentioned above?

**Comment by vineetravi107** [comment]
Yes, I have tried the above mentioned links. It is not working.
 Screenshot below:

![code](https://user-images.githubusercontent.com/9457234/29823580-b2b74476-8cec-11e7-8cdf-2b2916a0d0f0.JPG)

 I'm still getting the following error below:
![error](https://user-images.githubusercontent.com/9457234/29823599-c5d7cb34-8cec-11e7-97fd-2aecc99bb35c.JPG)


Is it the problem with 
"parameters.add("sslContextFactory", "org.restlet.ext.ssl.PkixSslContextFactory");" in PKRestServer.java file ??

**Comment by drallgood** [comment]
Have you tried IbmX509 instead of SunX509?

**Comment by vineetravi107** [comment]
No. I ll try it by tomorrow and let you know. I'm also willing to communicate in email for the same.

**Comment by vineetravi107** [comment]
This can be a simillar issue as https://github.com/mwkirk/javapns/issues/41.

**Comment by vineetravi107** [comment]
No.. its not working with IbmX509 either.

**Comment by drallgood** [comment]
Did you try IbmPKIX as suggested by https://github.com/mwkirk/javapns/issues/41#issuecomment-14411404

**Comment by vineetravi107** [comment]
Series <parameter> = ... line in setupSSL() method is creating problem.. when I'm calling this line, only the SUNX509 exception is coming.. I think the call is going to Java.security module internally which is causing this error.. Is there any code to create the same method i.e. setupSSL() using javax. jars??

**Comment by vineetravi107** [comment]
Can you provide alternative code for setting up SSL with javax jar.

**Comment by vineetravi107** [comment]
I did try IBMPKIX from https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.security.component.80.doc/security-component/jsse2Docs/pkixtrustmgrsupport.html 

I wanted a similar code supporting what you have written in steupSSL() method.

**Comment by stepio** [comment]
@vineetravi107 

You're referring to the code of example server, not to the code of Wallet API.
I'd recommend to disable SSL at all and configure it in ngnix, using nginx as a reverse proxy for your service.

**Comment by vineetravi107** [comment]
I'll try that.. in the mean time can I replace my setupSSL() method with accepted code here: https://stackoverflow.com/questions/8947148/java-ssl-code-throwing-nosuchalgorithexception

**Comment by vineetravi107** [comment]
I mean to say that will the following code do exact same thing what setupSSL() method do: 
`KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
InputStream inputStream = new FileInputStream("myTrustStore.jks");
trustStore.load(inputStream, "password".toCharArray());
TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
trustManagerFactory.init(trustStore);
TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
SSLContext sslContext = SSLContext.getInstance("SSL");
sslContext.init(null, trustManagers, null);`

**Comment by stepio** [comment]
I mean that you should not invoke method `setupSSL()`. I assume that you got it from here:
https://github.com/drallgood/jpasskit/blob/master/jpasskit.server/src/main/java/de/brendamour/jpasskit/server/PKRestServer.java
Then check next property:
```
		boolean useSSL = Boolean.parseBoolean(serverConfigurationProperties.getProperty(SERVER_BIND_SSL_ENABLED_KEY));
```
It should be set to `false`, then server will start without SSL and you'll configure SSL elsewhere.

**Comment by stepio** [comment]
Please note that SSL is still mandatory - iOS device won't be able to access this API without SSL, so you DO HAVE to configure it somewhere (in `nginx`, for example).

**Comment by vineetravi107** [comment]
Yes. I have set it to false and correct iOS device is unable to access this API. But code works properly and I get service hits using PostMan..
 Can you suggest some links on how to configure nginx in JAVA. I havent used this and the links that I'm looking into is not useful..

**Comment by stepio** [comment]
It's not in Java, `nginx` is a separate standalone service. Use Google please, your question is completely irrelevant for this project.

**Comment by vineetravi107** [comment]
Ok .. I'll do more research in this and try to implement.. thanks

**Comment by vineetravi107** [comment]
nginx would not ne feasible in my case due to immediate requirements. Can you provide any alternative to replace the following lines for code.
`Series<Parameter> parameters = restTrustedServer.getContext().getParameters();
		parameters.add("sslContextFactory", "org.restlet.engine.ssl.DefaultSslContextFactory");
		parameters.add("keystorePath", keystorePath);
		parameters.add("keystorePassword", keystorePassword);
		parameters.add("keyPassword", keyPassword);
		parameters.add("keystoreType", keystoreType);`

**Comment by drallgood** [comment]
@vineetravi107 I would have suggested exactly what @stepio did: disable SSL and use a reverse proxy like nginx or Apache. This is the recommended setup for production systems anyway.

If that doesn’t work for you for some reason or another, then I’m sorry to say I don’t think we’re able to help you. None of us knows anything about AIX.

If you can’t figure this out yourself, you essentially have two options:
- Hire someone who can help you with your issue
- Implement your own passkit server. Apple has a pretty good documentation that should help you with this task and you can use the jPasskit server for reference as well.

**Comment by vineetravi107** [comment]
I was able to resolve this issue. As I mentioned in my previous discussion that I was using IBM AIX OS which uses IBMX509, which is not supported by org.restlet framework. I came to know that `org.restlet.engine.ssl.DefaultSslContextFactory` internally used SUNX509 as KeyManagerFactory. I found this by decompiling DefaultSslContextFactory.java file. I changed all SUNX509 to IBMX509 and it worked.
